kernel/src/hil/hasher.rs

```diff
@@ -0,0 +1,83 @@
+//! Interface for Hasher
+
+use crate::utilities::leasable_buffer::LeasableBuffer;
+use crate::ErrorCode;
+
+/// Implement this trait and use `set_client()` in order to receive callbacks.
+///
+/// 'L' is the length of the 'u8' array to store the hash output.
+pub trait Client<'a, const L: usize> {
+    /// This callback is called when the data has been added to the hash
+    /// engine.
+    /// On error or success `data` will contain a reference to the original
+    /// data supplied to `add_data()`.
+    fn add_data_done(&'a self, result: Result<(), ErrorCode>, data: &'static mut [u8]);
+
+    /// This callback is called when a hash is computed.
+    /// On error or success `hash` will contain a reference to the original
+    /// data supplied to `run()`.
+    fn hash_done(&'a self, result: Result<(), ErrorCode>, hash: &'static mut [u8; L]);
+}
+
+/// Computes a non-cryptographic hash over data
+///
+/// 'L' is the length of the 'u8' array to store the hash output.
+pub trait Hasher<'a, const L: usize> {
+    /// Set the client instance which will receive `hash_done()` and
+    /// `add_data_done()` callbacks.
+    /// This callback is called when the data has been added to the hash
+    /// engine.
+    /// The callback should follow the `Client` `add_data_done` callback.
+    fn set_client(&'a self, client: &'a dyn Client<'a, L>);
+
+    /// Add data to the hash block. This is the data that will be used
+    /// for the hash function.
+    /// Returns the number of bytes parsed on success
+    /// There is no guarantee the data has been written until the `add_data_done()`
+    /// callback is fired.
+    /// On error the return value will contain a return code and the original data
+    fn add_data(
+        &self,
+        data: LeasableBuffer<'static, u8>,
+    ) -> Result<usize, (ErrorCode, &'static mut [u8])>;
+
+    /// Request the hardware block to generate a Hasher and stores the returned
+    /// hash in the memory location specified.
+    /// This doesn't return any data, instead the client needs to have
+    /// set a `hash_done` handler to determine when this is complete.
+    /// On error the return value will contain a return code and the original data
+    /// If there is data from the `add_data()` command asyncrously waiting to
+    /// be written it will be written before the operation starts.
+    ///
+    /// If an appropriate `set_mode*()` wasn't called before this function the
+    /// implementation should try to use a default option. In the case where
+    /// there is only one hash supported this should be used. If there is no
+    /// suitable or obvious default option, the implementation can return an
+    /// error with error code ENOSUPPORT.
+    fn run(&'a self, hash: &'static mut [u8; L]) -> Result<(), (ErrorCode, &'static mut [u8; L])>;
```

Comment:
Should there be a method to query the correct size of the hash? E.g. for SipHash this would be 8 or 16 bytes, but for other hashes it might be different.

Reply:
I _think_ that this is encoded in the const generic `L`, which you should be able to access given you have access to the precise `T: Hasher<L>` or a `&dyn Hasher<L>`.

Reply:
Right! Sorry, got confused.